### ISSUE TITLE
Missing pointer and possible buffer overrun in setup for PMM

### ISSUE THREAD

**Issue by bSchnepp** [bSchnepp/Feral#11]
Commit 17e9d1dfe27781fe55974aee1e648edddf71697d introduces a new PMM.
Under kern/mm/createpage.c, there is a possibility that a *very* big array is allocated (~200MB!!!) on the stack.

The kernel's stack is only 16KiB. This doesn't seem to be causing a fault (somehow) right now, but this shouldn't be possible.

```c
	UINT8 ShiftAmt = 3;	/* Map 1 bit to 8 bytes at least (2^3) */
	while (sz >>= 1)
	{ 
		++ShiftAmt;
	}
	
	FreeMemory >>= ShiftAmt;
	
	UINT8 map[FreeMemory];
```

In the case of 6GiB of free memory, this would be FreeMemory = (6 * 1024 * 1024 * 1024) / (8 * 4096), which is many many many times larger than 16384. Naturally, there should be a stack overrun expected here to deal with.

**Comment by bSchnepp** [comment]
Looks like it's definitely a buffer overrun.
```c
	pAllocInfo.FreeAreaRangeCount = FreeMemCount;
	MmFreeAreaRange ranges[FreeMemCount];
	for (UINT64 i = 0; i < FreeMemCount; ++i)
	{
		ranges[i].Start = FreeMemLocs[i];
		ranges[i].End = FreeMemLocs[i+1];
		KiPrintFmt("Start: %x\n", ranges[i].Start);
	}
```

In kernel_main.c causes behavior to change drastically.
Possible solutions are to use the first chunk of free memory we can get (which isn't a great idea, it's very small and hardware might want to DMA from there), or use some area in the largest contiguous free area.

Also consider using a different data structure, but we need O(1) access and O(1) for any changes, with the added constraint of being able to allocate at specific addresses of a specific size >= 0.

(long term solution is, of course, connecting up the security monitor and actually checking sType for all data structures, if invalid, bugcheck the kernel.)

**Comment by bSchnepp** [comment]
7fac409e33555299b15b409826f574050c23fdd4 provides a kernel panic if this issue is still present.

**Comment by bSchnepp** [comment]
Upon more careful examination of the issue, it's not a buffer overrun after all.
In kernel_main.c
```c
		MmFreeAreaRange ranges[FreeMemCount];
		for (UINT64 i = 0; i < FreeMemCount; i++)
		{
			ranges[i].sType = MM_STRUCTURE_TYPE_FREE_AREA_RANGE;
			ranges[i].pNext = (void*)(0);
			ranges[i].Start = FreeMemLocs[i];
			ranges[i].End = FreeMemLocs[i+1];
		}
		allocInfo.Ranges = ranges;
		MmCreateInfo info;

```

Ranges was never set, so a new array was declared with random garbage data, which looked like a buffer overrun.

Issue will be kept open since there's still going to be a problem in createpage.c with actually creating the bitmap.

**Comment by bSchnepp** [comment]
The new method of simply placing at whatever first area we can get works for now, but care should be done to make sure this is actually correct (ie, are we getting out of our boundary somehow?)

Should check to ensure boundaries are met and we don't overrun it.

**Comment by bSchnepp** [comment]
Issue should be fully resolved with b79bb6b5096caea853e3f76dcfbc8fb68647b078.
